### PR TITLE
OCP-2333 Add black, flake8 and mypy

### PR DESCRIPTION
**Pull request by pamarcos**
Add more and better formatters, linters and static analyzers to make sure we catch hairy issues ASAP.

I had to upgrade to Python 3.7 as the minimum required because of methods in a class returning a new instance of their own type. The `T = TypeVar('T', bound='TestCase')` is uglier and doesn't work well with mypy. More info at https://stackoverflow.com/questions/44640479/mypy-annotation-for-classmethod-returning-instance#comment119619794_44644576

### CODE REVIEW

**Comment by rgonzalezfluendo** [comment]
offtopic:  Changed in version 3.10: Unions can now be written as `X | Y`.

File: fluster/fluster.py [hunk 1]
```diff
@@ -55,10 +55,10 @@ def __init__(
         reference: bool = False,
         summary: bool = False,
         keep_files: bool = False,
-        threshold: int = None,
-        time_threshold: int = None,
+        threshold: Union[int, None] = None,
```
Comment (rgonzalezfluendo) [inline fluster/fluster.py:58]: You can use `Optional[int]` as a shorthand for `Union[int, None]`.

**Comment by rgonzalezfluendo** [comment]
LGTM. I prefer `Optional[X]` than `Union[X, None]`

**Review by rgonzalezfluendo — APPROVED** [review]

File: fluster/fluster.py (hunk 1 above)
Comment (pamarcos) [reply fluster/fluster.py]: Totally agree. I did a fixup: https://github.com/fluendo/fluster/pull/81/commits/d77322c8990c9e8ddd0bf70b96b92d7eb4a536ae

Squash + merge?

**Pull request merged**